core/src/main/java/fr/sncf/osrd/envelope_sim/allowances/MarecoAllowance.java

```diff
@@ -99,87 +111,178 @@ private DoubleArrayList findStops(Envelope base) {
+     * This process ensures the continuity of the final envelope.
      */
-    private void applyAllowanceRegion(EnvelopeBuilder builder, Envelope allowanceRegion, double addedTime) {
-        var baseTime = allowanceRegion.getTotalTime();
-        var baseDistance = allowanceRegion.getTotalDistance();
+    private void applyAllowanceRegion(EnvelopeBuilder builder, Envelope envelopeRegion) {
+
+        // build an array of the imposed speeds between ranges
+        // every time a range is computed, the imposed left and right speeds are memorized
+        var imposedTransitionSpeeds = new double[ranges.size() + 1];
+        imposedTransitionSpeeds[0] = envelopeRegion.getBeginSpeed();
+        for (int i = 1; i < ranges.size(); i++)
+            imposedTransitionSpeeds[i] = NaN;
+        imposedTransitionSpeeds[ranges.size()] = envelopeRegion.getEndSpeed();
+
+        // build an array of (range, percentage) in order to sort the array rangeOrder by ascending percentage
+        var rangePercentages = new RangePercentage[ranges.size()];
+        for (var i = 0; i < ranges.size(); i++) {
+            var range = ranges.get(i);
+            var percentage = range.value.getAllowanceRatio(
+                    envelopeRegion.getTimeBetween(range.beginPos, range.endPos),
+                    range.beginPos - range.endPos
+            );
+            rangePercentages[i] = new RangePercentage(range, percentage);
+        }
 
-        // TODO: compute the slowest possible trip, and ensure the target time is within the realms of possibility
+        // the order in which the ranges should be computed
+        // ranges are computed with increasing percentage values
+        var rangeOrder = new Integer[ranges.size()];
+        for (var i = 0; i < ranges.size(); i++)
+            rangeOrder[i] = i;
+        Arrays.sort(rangeOrder, Comparator.comparingDouble(rangeIndex -> rangePercentages[rangeIndex].percentage));
+
+        var res = new Envelope[ranges.size()];
+
+        // compute ranges one by one in the right order
+        for (var rangeIndex : rangeOrder) {
+            var range = ranges.get(rangeIndex);
+            logger.debug("computing range n°{}", rangeIndex + 1);
+            var envelopeRange = Envelope.make(envelopeRegion.slice(range.beginPos, range.endPos));
+            var imposedBeginSpeed = imposedTransitionSpeeds[rangeIndex];
+            var imposedEndSpeed = imposedTransitionSpeeds[rangeIndex + 1];
+            var allowanceRange =
+                    computeAllowanceRange(envelopeRange, range.value, imposedBeginSpeed, imposedEndSpeed);
+            // memorize the begin and end speeds
+            imposedTransitionSpeeds[rangeIndex] = allowanceRange.getBeginSpeed();
+            imposedTransitionSpeeds[rangeIndex + 1] = allowanceRange.getEndSpeed();
+            res[rangeIndex] = allowanceRange;
+        }
+
+        for (var envelope : res) {
+            builder.addEnvelope(envelope);
+        }
```

Comment:
I think it makes slightly more sense to return res, remove the builder argument, and do that in the caller

Reply:
Done